rfcs/0006-transaction.md

```diff
@@ -0,0 +1,300 @@
+
+The conflict checking in Badger is as follows:
+
+1. Track the sequence number when transaction started.
+2. Track both read and write keys in the transaction state.
+3. When `commit()` is called, check conflicts with the recent committed transactions in the Oracle which committed after the transaction started.
+4. The conflict check verifies that: the transaction's read and write keys do not conflict with other transactions' write keys
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn_opts = TransactionOptions::new().with_isolation_level(IsolationLevel::Serializable); // or IsolationLevel::Snapshot
+let txn = kv_store.transaction_with_opts(txn_opts).await.unwrap();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
```

Comment:
how do txns play with ttl/time? my head hurts a little bit thinking about this... do non-committed transactions progress time? Here's what I'm worried about.

Imagine (a) expires at time=3

| time | `txn1`          | `txn2`         | `txn3` |
|------|-----------------|----------------|--------|
| 0    | vA = read(a)    |                | vD = read(d)       |    
| 4    |                 | write(b, vB)   |                    |
| 5    |                 |                | vA = read(a) |
| 6    |                 | commit()       |        |
| 7    | write(c, vA)    |                | write(e, vD + va)       |
| 8    | commit()        |                |        |

The possible problems:
1. `txn1` read vA when `a` was still alive, but writes the value to c after vA is expired because `txn2` progressed time by writing `b` at t=4
2. `txn3` starts at t=0, at which point `a` is still alive, no committed operation has yet been committed when it reads `a` at t=5, but `txn2` has progressed time so that a is now expired. if `txn2` is cancelled, time will have progressed due to a zombie transaction

Not sure what we want the behavior of time progression with regards to uncommitted transactions to be... Perhaps time only progresses on a transaction commit? That means that situation 1 above is OK because the txn started before time progressed, whereas situation 2 would not happen because time doesn't progress until txn2 commits - so if txn2 aborts time would not have progressed

(NOTE: right now time only progresses on writes, but I think after we implement #322 we'll have time progress on reads as well)

Reply:
After discussion with @hachikuji and @rodesai we came up with a pretty neat solution to this problem - if we track the earliest expire_ts from any key that we read, the txn coordinator will consider a conflict if any other commit causes the timestamp to advance past that earliest expire_ts. Time only advances at a commit.

Example: (`valA` expire_ts=3)
| time | txn1 | txn2 |
|------|-------|------|
| 0 | valA = read(A) | |
| 4 | | write(b) |
| 5 | | commit() |
| 6 | write(C, valA) | |
| 7 | commit | |

The commit at time=7 fails because valA expire_ts = 3 which is less than the committed time of 5.

Reply:
I forgot to elaborate on the most important note, which is that the act of progressing time actually functions as a write on potentially a massive keyspace (all keys that expired during that time) so we need to track those conflicts in the oracle.

Reply:
> I actually think we should leave time only progressing on writes

i feel we can reference what badger does here, it also contains a `expire_ts` support. as far as i know, the `expire_ts` seems not related with making snapshot & conflict checking at all. 👀

all it does is simply made a check with `isDeletedOrExpired` during `get()` at https://github.com/dgraph-io/badger/blob/main/txn.go#L486, and `isDeletedOrExpired` is simply a comparing with system clock:

```
func isDeletedOrExpired(meta byte, expiresAt uint64) bool {
	if meta&bitDelete > 0 {
		return true
	}
	if expiresAt == 0 {
		return false
	}
	return expiresAt <= uint64(time.Now().Unix())
}
```

as it looks, it did not try to make the read with `expireAt` consistent during a transaction, if we read a key with `expireAt` twice in a transaction, it's possible to get a `Some(val)` first, but get a `None` later. 🤔

Reply:
Is there any reason not to track the `min(expire_ts)` for all rows read and consider it a conflict if another transaction causes time to progress past that? It feels like Badger has a bug here 😅 seems like a simple fix on our end because db time is enforced monotonic and only progresses on writes (unlike badger, which uses system clocks)

Reply:
This is a great point.
With TTL, if the transaction started before the key expired, but committed after the key expired, then the `expiry` of the key must be treated as a `write` to the key, even if no other transactions happened.
The transaction itself will have a timestamp, and all the keys that were read must be valid as of this timestamp.

Later on, I see that `expire_time` is also tracked for every transaction. I think this is set when the transaction starts. If so, we already know which keys could potentially expire and perhaps track the expire time for those keys in the transaction state.

Reply:
I think time progression should not depend on whether there was another transaction or not, time should be retrieved when the transaction commits.

Reply:
> Is there any reason not to track the min(expire_ts) for all rows read and consider it a conflict if another transaction causes time to progress past that? It feels like Badger has a bug here 😅 

i guess there're few use case really depend on this timing semantic? if no one raise bug for them, i think it's not bad to keep things simple & stupid. 👀

when we have `expire_ts`, it's likely to be regard something as a cache imo. and if it's a cache, it's better to assume this key/value could be removed by others near the expiration time in the program, and we can have a logic to rebuild this row if this row is expired.

can you help offer some examples on real world use cases when we need this deterministic on this expiration time in transaction?

Reply:
That's fair, let's add this to the rejected alternatives section so that we remember the discussion if we do ever want to add it back. The use cases I was worried about are when you are processing historical data (e.g. in stream processing this happens pretty frequently) and you are advancing db-time much faster than wall-clock time -- you can process a month of historical data in a few hours. I can't think of a specific use where you would also want to do transactions.

Reply:
> so that we remember the discussion if we do ever want to add it back.

+1 I definitely feel that we should come back to this. If we don't account for time when doing conflict detection we're certainly breaking SSI. A bit pedantic, but I think it would be more appropriate to add a future work section that documents it as a known limitation and references this discussion as a proposed solution.

Reply:
I was thinking on this some more. Are we saying that time won't advance _within_ a transaction until commit Suppose I have:

```
t0, begin
t1, put(k1, v1, ttl=t3)
t2, get(k1) # returns v1
t3, ...
t4, get(k1) # what does this return?
tN, commit
```

Or are we saying that time _does_ advance within a transaction, but _only_ for reads in the transaction. The global/external/orcale time isn't advanced until commit?

Reply:
This is a really good point. What do you think about an entire transaction taking in place as a single, atomic operation? This means that time doesn't advance within a transaction at all (reads included) until a commit happens.

That means that the situation you outlined where each operation has a timestamp would actually look more like:
```
t0, begin, grab clock tick -> t1
t1, put(k1, v1, ttl=3)
t1, get(k1) # returns v1
t1, ...
t1, get(k1) # returns v1
t1 commit
```

This is within the time semantics of a clock being monotonically increasing, but allowed to return the same time so long as it doesn't go backwards in time. 

Technically, it's possible that any operation that advances time would be considered a conflict but conflicts only manifest in reading TTL at the moment so we need to track and see whether a conflict "really" happened (by tracking the TTL of all rows read).

Reply:
I suppose another approach we could consider might be making expiration a real write operation in background. It would have clear semantics on SSI, at the expense of expiration happening later. 🤔

What I'm worried about is whether the scope of conflict checking will become bloated as features increase, and it will put more burden on correctness proof as more features increase. In contrast, checking only the sequence numbers in conflict checking is well understood in the industry, not only LSM engines but also very common in the BTree family.

However, if we can name some existing projects that take this approach on conflict checking for TTL, we can easily follow that path.